File: apps/ios/Sources/Core/Realtime/NativeRealtimeManager.swift
import Foundation
import Combine

// The SDK bridge exposes only read-only callbacks, never application emit/join.
@MainActor protocol RealtimeSocketDriver: AnyObject {
    func onConnected(_ callback: @escaping @MainActor () -> Void)
    func onDisconnected(_ callback: @escaping @MainActor () -> Void)
    func onError(_ callback: @escaping @MainActor () -> Void)
    func onSyncRequired(_ callback: @escaping @MainActor (String) -> Void)
    func connect()
    func removeAllHandlers()
    func disconnect()
}
@MainActor protocol RealtimeSocketFactory {
    func create(options: RealtimeConnectionOptions) throws -> any RealtimeSocketDriver
}
enum RealtimeEvent: Sendable {
    case syncRequired(RealtimeScope), revalidationRequired(RealtimeTicket)
}

// Modified reuse of Meloming SongLiveSocketManager: retained driver, eventSubject,
// weak callback capture, connect/setupEventHandlers/disconnect/reconnect ownership.
// No join/identifier/raw event payload. Every reconnect needs current REST authority.
@MainActor final class NativeRealtimeManager: ObservableObject {
    @Published private(set) var connectionStatus = RealtimeStatus.disconnected
    private var socket: (any RealtimeSocketDriver)?
    private let factory: any RealtimeSocketFactory
    private var lifecycle = RealtimeLifecycle()
    private let eventSubject = PassthroughSubject<RealtimeEvent, Never>()
    var events: AnyPublisher<RealtimeEvent, Never> { eventSubject.eraseToAnyPublisher() }
    init(factory: any RealtimeSocketFactory) { self.factory = factory }
    func bind(scope: RealtimeScope?, foreground: Bool) {
        if lifecycle.bind(scope: scope, foreground: foreground) { disposeSocket(); connectionStatus = lifecycle.status }
    }
    func connect(scope: RealtimeScope, bearer: String) {
        guard let ticket = lifecycle.begin(expected: scope) else { return }
        disposeSocket(); connectionStatus = lifecycle.status
        do {
            let socket = try factory.create(options: RealtimeConnectionOptions(scope: scope, bearer: bearer))
            self.socket = socket
            setupEventHandlers(socket: socket, ticket: ticket)
            socket.connect()
        } catch { lost(ticket) }
    }
    func reconnect(ticket: RealtimeTicket, bearer: String) {
        guard lifecycle.canReconnect(ticket) else { return }
        connect(scope: ticket.scope, bearer: bearer)
    }
    func disconnect() { bind(scope: nil, foreground: false) }
    private func disposeSocket() {
        let previous = socket; socket = nil
        previous?.removeAllHandlers(); previous?.disconnect()
    }
    private func setupEventHandlers(socket: any RealtimeSocketDriver, ticket: RealtimeTicket) {
        socket.onConnected { [weak self] in
            guard let self, lifecycle.connected(ticket) else { return }
            connectionStatus = lifecycle.status
            eventSubject.send(.syncRequired(ticket.scope))
        }
        socket.onDisconnected { [weak self] in self?.lost(ticket) }
        socket.onError { [weak self] in self?.lost(ticket) }
        socket.onSyncRequired { [weak self] payload in
            guard let self, lifecycle.wake(ticket, payload: payload) else { return }
            eventSubject.send(.syncRequired(ticket.scope))
        }
    }
    private func lost(_ ticket: RealtimeTicket) {
        guard lifecycle.disconnected(ticket) else { return }
        disposeSocket(); connectionStatus = lifecycle.status
        eventSubject.send(.revalidationRequired(ticket))
    }
    isolated deinit {
        socket?.removeAllHandlers()
        socket?.disconnect()
    }
}

File: apps/android/app/src/main/java/chat/rogi/rogichat/core/realtime/NativeRealtimeManager.kt
package chat.rogi.rogichat.core.realtime

import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow

/** Required production Socket.IO bridge; deliberately has no application emit API. */
interface RealtimeSocketDriver {
    fun onConnected(callback: () -> Unit)
    fun onDisconnected(callback: () -> Unit)
    fun onError(callback: () -> Unit)
    fun onSyncRequired(callback: (String) -> Unit)
    fun connect()
    fun removeAllHandlers()
    fun disconnect()
}
fun interface RealtimeSocketFactory { fun create(options: RealtimeConnectionOptions): RealtimeSocketDriver }
sealed interface RealtimeEvent {
    data class SyncRequired(val scope: RealtimeScope) : RealtimeEvent
    data class RevalidationRequired(val ticket: RealtimeTicket) : RealtimeEvent
}

/** Modified reuse of Meloming SongLiveSocketManager's callbackFlow/on/connect/
 * awaitClose teardown. Removes all join/application emits and logs; binds every
 * callback to an immutable operation and the owner's persistent session epoch.
 */
class NativeRealtimeManager(private val factory: RealtimeSocketFactory) {
    private val lock = Any()
    private val lifecycle = RealtimeLifecycle()
    private class Connection(val ticket: RealtimeTicket, val driver: RealtimeSocketDriver, val closeFlow: () -> Unit) {
        fun dispose() { driver.removeAllHandlers(); driver.disconnect() }
    }
    private var active: Connection? = null
    val status: RealtimeStatus get() = synchronized(lock) { lifecycle.status }

    /** Call synchronously before session/logout/background async work. */
    fun bind(scope: RealtimeScope?, foreground: Boolean) = synchronized(lock) {
        if (lifecycle.bind(scope, foreground)) {
            val previous = active; active = null
            previous?.dispose(); previous?.closeFlow()
        }
    }
    fun connect(scope: RealtimeScope, bearer: String): Flow<RealtimeEvent> = open(scope, bearer, null)
    /** Owner first awaits real REST authorization, then passes the same reconnect ticket. */
    fun reconnect(ticket: RealtimeTicket, bearer: String): Flow<RealtimeEvent> = open(ticket.scope, bearer, ticket)

    private fun open(scope: RealtimeScope, bearer: String, reconnect: RealtimeTicket?): Flow<RealtimeEvent> = callbackFlow {
        var connection: Connection? = null
        synchronized(lock) {
            if (reconnect != null && !lifecycle.canReconnect(reconnect)) { close(); return@synchronized }
            val ticket = lifecycle.begin(scope)
            if (ticket == null) { close(); return@synchronized }
            val previous = active; active = null
            previous?.dispose(); previous?.closeFlow()
            try {
                val socket = factory.create(RealtimeConnectionOptions(scope, bearer))
                val current = Connection(ticket, socket) { close() }
                active = current; connection = current
                fun lost() = synchronized(lock) {
                    if (lifecycle.disconnected(ticket)) {
                        current.dispose()
                        trySend(RealtimeEvent.RevalidationRequired(ticket))
                        close()
                    }
                }
                socket.onConnected {
                    synchronized(lock) {
                        if (lifecycle.connected(ticket)) trySend(RealtimeEvent.SyncRequired(scope))
                    }
                }
                socket.onDisconnected { lost() }
                socket.onError { lost() }
                socket.onSyncRequired { payload ->
                    synchronized(lock) {
                        if (lifecycle.wake(ticket, payload)) trySend(RealtimeEvent.SyncRequired(scope))
                    }
                }
                socket.connect()
            } catch (_: Exception) {
                // Driver failures never expose bearer-bearing configuration or raw errors.
                connection?.dispose()
                if (lifecycle.disconnected(ticket)) trySend(RealtimeEvent.RevalidationRequired(ticket))
                close()
            }
        }
        awaitClose {
            synchronized(lock) {
                val current = connection
                if (current != null) {
                    if (active === current) {
                        active = null
                        if (!lifecycle.canReconnect(current.ticket)) lifecycle.close(current.ticket)
                    }
                    current.dispose()
                }
            }
        }
    }
}
